### ISSUE TITLE
Cannot use aggregations with numbers stored as strings in MySQL in v50

### ISSUE THREAD

**Issue by martin-knap** [metabase/metabase#44431]
### Describe the bug

Hello,
Since the v50 i am getting errors when trying to sum values/dates saved as string in db (mysql/snowflake) even though i changed table metadata to quantity. It was working fine before now i get this error:
`Invalid query: {:stages [{:aggregation [[nil nil [nil nil ["expression returning a number"]]]]}]}`
`Invalid query: {:stages [{:filters [nil [nil nil ["expression returning a date, time, or date time"]]]}]}`
### To Reproduce

1. data (numbers) saved as varchar in database table (mysql or snowflake)
2. set table metadata in metabase admin to quantity for the given field
3. trying to sum/aggregate values (icon remains as [T] before field name in UI when selecting the field)
![CleanShot 2024-06-19 at 19 52 25@2x](https://github.com/metabase/metabase/assets/96686504/b1469b1a-a872-49fc-96dd-d30de21106d9)
![CleanShot 2024-06-19 at 19 54 41@2x](https://github.com/metabase/metabase/assets/96686504/d0aceb4d-431a-40f2-94e2-e969e0a88202)
![CleanShot 2024-06-19 at 19 55 22@2x](https://github.com/metabase/metabase/assets/96686504/e053f407-e72c-446a-948d-c3cb3036a8c3)

### Expected behavior

sum of text fields set to number in metabase metadata worked fine in v.0.49.x

### Information about your Metabase installation

```JSON
metabase v0.50.6
```

### Severity

high

**Comment by paoliniluis** [comment]
This implies that you were able to sum strings before?

**Comment by martin-knap** [comment]
yes it worked when you changed type in table metadata in admin or in model

**Comment by paoliniluis** [comment]
maybe I'm not seeing something but why do you save numbers as strings in the database?

**Comment by martin-knap** [comment]
some ETL tools (e.g. Keboola) save all data in Snowflake as varchar and i cant change the setting or update metadata in storage :(

**Comment by armetiz** [comment]
Same problem here after upgrade to latest metabase OSS release. Also have a VARCHAR

**Comment by armetiz** [comment]
I've solved the problem, but I'm not sure I'm dealing with exactly the same issue as the author.

The problem was still present after changing the column type from VARCHAR to FLOAT. 
I checked the metabase error-log, and observed several information messages about type conversions. So I checked the correspondence of my model. The field had no specific category, so I associated my field with a "price", after which it worked.

Regards.

**Comment by jkohlbach** [comment]
We tried the fix @armetiz suggested but it didn't work out. Probably because we kept the columns as varchar, I'm guessing.

This seems like a regression, we're also in a situation where the column needs to be varchar and this has broken a ton of custom expressions. You can no longer sum or sumif for example.

Any chance this will be fixed on the next release or should we downgrade?

**Comment by salman2learn** [comment]
Same issue with OSS version using sqlite number data type.
Thanks.

**Comment by SeshTiliRest** [comment]
I have a similar problem with my multiple replacement after updating from v0.49.14 to v0.50.18.

![invalid](https://github.com/user-attachments/assets/5b6af0b7-36b2-4c98-891b-2bb272baa9d5)
`replace(replace(replace([Accept], "1", "А"), "0", "В"), "2", "К")`

![problem](https://github.com/user-attachments/assets/38037bc2-1aec-47fb-a66f-9795a9066855)
`Invalid query: {:stages [{:expressions [[nil nil [nil nil [nil nil ["expression returning a string"]]]]]}]}`

After deleting custom column with dat replace function - error is gone.

**Comment by douglas-regatieri** [comment]
I encountered a similar error after updating to v0.50. I'm using Redshift, and the issue occurs with a SUM on a float8 datatype

**Comment by martin-knap** [comment]
this is also affecting columns formatted as number and trying to do `substring([Date Key], 1, 4)`

**Comment by Gourds** [comment]
+1

**Comment by musojames** [comment]
I've had the same issue with BigQuery. Downgrading back to 49 has unblocked existing charts for me. Note that you need to run the migrations downgrade command with version 50, before launching 49 again to rollback.

**Comment by musojames** [comment]
I did it running in docker. If you have an environment where you can connect to the DB directly run 
``` 
docker run \
    -it \
    --entrypoint "/bin/sh" \
    -e MB_DB_CONNECTION_URI ="db connection string here" \
    -e MB_ENCRYPTION_SECRET_KEY = "Your encryption secret if you set one" \
    metabase/metabase:v0.50.X # X here is the version you deployed. I had issues when my bug fix version didn't match the deployed.
```

this will get you a Metabase container running with a shell session. 

Once in the shell session run
``` 
cd /app
java -jar metabase.jar migrate down
```` 

Once you have successfully run the migration down, then deploy a `v0.49.*` version instead. 


If you can't get direct access to your metabase DB. You can do the same as above by deploying a container in a location where it does have access to the db, with a script that does the above as the entrypoint.

**Comment by paoliniluis** [comment]
Here's the answer of our lead architect around this issue as it sees pretty common now:
```
We never intentionally supported this, we improved expression validation a lot in MLv2 which is probably why it errors now whereas the old code just ran it with no idea if it made sense or not or if it would even work
```

I'm going to change this issue into an official feature request to coerce strings to integers

**Comment by paoliniluis** [comment]
@MarleTangible please create a separate issue
